src/universe.rs

```diff
@@ -1446,79 +1451,163 @@ impl Universe {
+        //    Ok(None).
         let mut opt_gen0_idx = None;
+        let mut opt_largest_gen_idx = None;
+        let mut opt_largest_gen_value = None;
         for i in 0..self.gen_states.len() {
             if let Some(gen) = self.gen_states[i].gen_or_none {
-                if gen == diff.gen0 {
-                    opt_gen0_idx = Some(i);
+                if diff.gen0 > 0 {
+                    if gen == diff.gen0 {
+                        opt_gen0_idx = Some(i);
+                    }
+                }
+                if let Some(largest_gen) = opt_largest_gen_value {
+                    if gen > largest_gen {
+                        opt_largest_gen_value = Some(gen);
+                        opt_largest_gen_idx   = Some(i);
+                    }
+                } else {
+                    opt_largest_gen_value = Some(gen);
+                    opt_largest_gen_idx   = Some(i);
                 }
                 // 2) ensure that no generation is >= gen1; if so, return Ok(None)
                 if gen >= diff.gen1 {
                     return Ok(None);
                 }
             }
         }
-        if opt_gen0_idx.is_none() {
+        if diff.gen0 > 0 && opt_gen0_idx.is_none() {
             return Ok(None);
         }
-        let gen0_idx = opt_gen0_idx.unwrap();
 
         // 3) make room for the new gen_state (make room in the circular buffer)
         for i in 0..self.gen_states.len() {
             if let Some(gen) = self.gen_states[i].gen_or_none {
-                if gen <= diff.gen1 - gen_state_len {
+                if gen + gen_state_len <= diff.gen1 {
                     self.gen_states[i].gen_or_none = None; // indicate uninitialized
                 }
             }
         }
 
-        // 4a) copy from the gen_state for gen0 to what will be the gen_state for gen1
+        let gen1_idx = if let Some(largest_gen_idx) = opt_largest_gen_idx {
+            let largest_gen = opt_largest_gen_value.unwrap();
+            (largest_gen_idx + diff.gen1 - largest_gen) % gen_state_len
+        } else {
+            0
+        };
+        // 4) If incremental update, then copy from the gen_state for gen0 to what will be the
+        //    gen_state for gen1.
         // TODO: This needs some serious cleanup
-        let gen1_idx = (gen0_idx + diff.gen1 - diff.gen0) % gen_state_len;
-        let cut_idx = cmp::max(gen0_idx, gen1_idx);
-        {
-            let (lower, upper): (&mut [GenState], &mut [GenState]) = self.gen_states.split_at_mut(cut_idx);
-            if gen1_idx < cut_idx {
-                lower[gen1_idx].clear();
-                upper[gen0_idx - cut_idx].copy(&mut lower[gen1_idx]); // this is an |= operation, hence the clear before this
-            } else {
-                upper[gen1_idx - cut_idx].clear();
-                lower[gen0_idx].copy(&mut upper[gen1_idx - cut_idx]); // this is an |= operation, hence the clear before this
+        if diff.gen0 > 0 {
```

Comment:
Could perhaps use `inline` functions to describe what this is doing for clean-up/readability.

Reply:
Yeah, I'd like to move this to a function. I felt a little dirty writing it like this...